File: 3748-sort-matrix-by-diagonals/sort-matrix-by-diagonals.go
func sortMatrix(grid [][]int) [][]int {
	n := len(grid)
	diagonals := make(map[int][]int)

	// Step 1: Collect numbers by diagonals (key = i - j)
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			key := i - j
			diagonals[key] = append(diagonals[key], grid[i][j])
		}
	}

	// Step 2: Sort each diagonal
	for key, diag := range diagonals {
		if key >= 0 {
			// bottom-left: sort descending
			sort.Slice(diag, func(a, b int) bool {
				return diag[a] > diag[b]
			})
		} else {
			// top-right: sort ascending
			sort.Ints(diag)
		}
		diagonals[key] = diag
	}

	// Step 3: Place back into grid
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			key := i - j
			grid[i][j] = diagonals[key][0]
			diagonals[key] = diagonals[key][1:] // pop front
		}
	}

	return grid
}

File: 3748-sort-matrix-by-diagonals/sort-matrix-by-diagonals.py
class Solution:
    def sortMatrix(self, grid: List[List[int]]) -> List[List[int]]:
        n = len(grid)
        diagonals = {}

        # Step 1: Collect numbers by diagonals (key = i - j)
        for i in range(n):
            for j in range(n):
                key = i - j
                if key not in diagonals:
                    diagonals[key] = []
                diagonals[key].append(grid[i][j])

        # Step 2: Sort each diagonal
        for key, diag in diagonals.items():
            if key >= 0:  # bottom-left (including main diagonal)
                diag.sort(reverse=True)  # descending
            else:  # top-right
                diag.sort()  # ascending

        # Step 3: Place back into grid
        for i in range(n):
            for j in range(n):
                key = i - j
                grid[i][j] = diagonals[key].pop(0)

        return grid
